embedded-graphics/src/primitives/circle.rs

```diff
@@ -206,37 +203,105 @@ where
             -(self.primitive.radius as i32),
             -(self.primitive.radius as i32),
         );
+
+        // A stroke width of zero renders a 1px stroke, so add 1 to inner radius to compensate
+        let inner_radius = self.primitive.radius as i32 - self.style.stroke_width_i32() + 1;
+        let outer_radius = self.primitive.radius as i32;
+
+        let inner_threshold = inner_radius.pow(2) - inner_radius;
+        let mut outer_threshold = outer_radius.pow(2) + outer_radius;
```

Comment:
Why is the `inner_radius` subtracted and the `outer_radius` added? IMO they should be treated equally and both be added. This way these examples should output the exact same results (with `r2 < r1`):
```rust
egcircle!(c, r1, fill_color = Some(Rgb888::RED));
egcircle!(c, r2, fill_color = Some(Rgb888::GREEN));
```
```rust
egcircle!(
    c, r1,
    fill_color = Some(Rgb888::GREEN),
    stroke_color = Some(Rgb888::RED),
    stroke_width = r1 - r2
);
```

Reply:
Honestly, because [this comment](https://stackoverflow.com/questions/1201200/fast-algorithm-for-drawing-filled-circles#comment80182898_1237519) recommended it. I didn't look into why it works any further than implementing it.

If I do change both to an add, I get this incorrect result:

![image](https://user-images.githubusercontent.com/800791/69831065-a8975e80-121f-11ea-8d77-b18fdf9eca09.png)

Reply:
With the original code, I think I get the correct result with `r1 = 20`, `r2 = 10`:

![image](https://user-images.githubusercontent.com/800791/69831257-894d0100-1220-11ea-96ed-a928cf97c0e6.png)

(ignore the on-screen text)

Am I missing something?

Reply:
I think this would be the mathematically correct version:
https://gist.github.com/rfuest/83bf0b06c82dec13449932adfc520863

This version uses 2 times larger length values to represent half pixel values.

The threshold calculation is derived from:
`(2(r + 0.5))^2 = (2r + 1)^2 = 4r^2 + 4r + 1`

Reply:
Ah that looks loads better. I hope you don't mind, I've shamelessly plugged that gist into this PR. Nice one figuring out the doubled solution.

Reply:
Oops, my code isn't perfect at all:
```rust
#[test]
fn stroke_width_doesnt_affect_fill() {
    let mut expected = MockDisplay::new();
    let mut style = PrimitiveStyle::with_fill(BinaryColor::On);
    Circle::new(Point::new(5, 5), 4)
        .into_styled(style)
        .draw(&mut expected);

    let mut with_stroke_width = MockDisplay::new();
    style.stroke_width = 1;
    Circle::new(Point::new(5, 5), 4)
        .into_styled(style)
        .draw(&mut with_stroke_width);

    assert_eq!(expected, with_stroke_width);
}
```

fixed version of `next`:
```rust
/ https://stackoverflow.com/a/1237519/383609
// https://stackoverflow.com/questions/1201200/fast-algorithm-for-drawing-filled-circles#comment80182898_1237519
fn next(&mut self) -> Option<Self::Item> {
    // If the fill and stroke colors are `None`, treat entire object as transparent and exit
    // early.
    if self.style.stroke_color.is_none() && self.style.fill_color.is_none() {
        return None;
    }

    loop {
        let len = (2 * self.p.x).pow(2) + (2 * self.p.y).pow(2);

        let color = if len < self.inner_threshold {
            self.style.fill_color
        } else if len < self.outer_threshold {
            // Use fill_color if no stroke_color was set
            self.style.stroke_color.or(self.style.fill_color)
        } else {
            None
        };
        let item = color.map(|c| Pixel(self.center + self.p, c));

        self.p.x += 1;

        if self.p.x > self.radius as i32 {
            self.p.x = -(self.radius as i32);
            self.p.y += 1;
        }

        if self.p.y > self.radius as i32 {
            break None;
        }

        if item.is_some() {
            break item;
        }
    }
}
```

Reply:
Fixed :slightly_smiling_face: test added in as well.